src/neuralnet/superblock.cpp

```diff
@@ -109,74 +311,226 @@ std::string PackBinarySuperblock(std::string sBlock)
 // Class: Superblock
 // -----------------------------------------------------------------------------
 
-void Superblock::LoadStats(const ScraperStats& stats)
+Superblock::Superblock()
+    : m_version(Superblock::CURRENT_VERSION)
+    , m_height(0)
+    , m_timestamp(0)
+{
+}
+
+Superblock::Superblock(uint32_t version)
+    : m_version(version)
+    , m_height(0)
+    , m_timestamp(0)
+{
+}
+
+Superblock Superblock::FromStats(const ScraperStats& stats)
 {
-    // TODO: this routine depends entirely on the order of items in the scraper
-    // stats map. Need to verify that it isn't too fragile...
-    for (const auto& entry: stats) {
+    // The loop below depends on the relative value of these enum types:
+    static_assert(statsobjecttype::NetworkWide < statsobjecttype::byCPID);
+    static_assert(statsobjecttype::byCPID < statsobjecttype::byCPIDbyProject);
+    static_assert(statsobjecttype::byProject < statsobjecttype::byCPIDbyProject);
+
+    Superblock superblock;
+
+    for (const auto& entry : stats) {
+        // A CPID, project name, or project name/CPID pair that identifies the
+        // current statistics record:
+        const std::string& object_id = entry.first.objectID;
+
         switch (entry.first.objecttype) {
+            case statsobjecttype::NetworkWide:
+                // This map starts with a single, network-wide statistics entry.
+                // Skip it because superblock objects will recalculate the stats
+                // as needed after deserialization.
+                //
+                continue;
+
             case statsobjecttype::byCPID:
-                {
-                    uint16_t mag = std::round(entry.second.statsvalue.dMag);
-
-                    if (mag > 0) {
-                        m_cpids.emplace(NN::Cpid::Parse(entry.first.objectID), mag);
-                    } else {
-                        m_network.m_zero_mag_cpid_count++;
-                    }
-                }
+                superblock.m_cpids.Add(
+                    Cpid::Parse(object_id),
+                    std::nearbyint(entry.second.statsvalue.dMag));
+
                 break;
 
             case statsobjecttype::byProject:
-                m_projects.emplace(entry.first.objectID, ProjectStats(
-                    std::round(entry.second.statsvalue.dAvgRAC),
-                    std::round(entry.second.statsvalue.dRAC)));
+                superblock.m_projects.Add(object_id, ProjectStats(
+                    std::nearbyint(entry.second.statsvalue.dAvgRAC),
```

Comment:
Per our discussion on Slack, change these back to std::round...

Reply:
✔️